### PR TITLE
Remade Vertex Cache

### PR DESCRIPTION
**Pull request by FernandoS27**
This one improves both versions and is more readable in general.

The new vertex cache simulates better how vertex caches work on GPUs. The size of it is 64Kb (aka 512 entries). Testing found this size produced almost no misses at all on most games. (where the original 32 entries cache produced many misses around the hundreds). Optimally speaking, the most optimal number should be between 256 or 1024 for the best trade off.

Please feel free to test it, I've left all the debugging code for tracking the cache misses (must be compiled on debug in msvc).

### CODE REVIEW

**Comment by tabnk** [comment]
New PR Retested - Build on GCC 6.3.0 with SSE4.1. On Windows 10 Pro.

Pokemon Sun - State 1
Master: 48 fps
New PR: 52 fps

Super Mario 3D Land - Initial Scene
Master: 60 fps
New PR: 62 fps

Mario Kart 7 - Mushroom Cup
Master: 29 fps
New PR: 30 fps

New Super Mario Bros 2 - 1-1 
Master: 77 fps
New  PR: 83 fps

Poochy And Yoshi s Woolly World - Opening Scene
Master: 50 fps
New PR: 50 fps

**Comment by wwylele** [comment]
>// TODO: when parallizing the shader units, this should be refactored into a
>// 'multiton' aka 4 singletons with it's own size.

This is the reason why I generally don't like your optimization at this point. You are dumping more and more TODOs to who wants to implement features, which leads more refactors and can cause bugs

**Comment by FernandoS27** [comment]
>.> that TODO had to be done either way if somebody ever wanted to parallize the shader units, or you want me to do that TODO.

File: src/video_core/vertex_cache.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+// stores positions that are cached;
+static std::array<u16, cache_size> remapper = {};
+
+// used for clearing the cache;
+u32 min_index = cache_size;
+u32 max_index = 0;
+} // Anonymous namespace
+
+inline bool contains(u32 position) {
```
Comment (wwylele) [inline src/video_core/vertex_cache.h]: Capitalize the first character `Contains`. So do other functions

File: src/video_core/vertex_cache.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include "common/common_types.h"
+#include "video_core/shader/shader.h"
+
+// This little module emulates PICA's vertex shader's post transform vertex cache.
```
Comment (wwylele) [inline src/video_core/vertex_cache.h]: Just a question from the word "emulates": does PICA actually have a vertex cache? And does it work like this code?

File: src/video_core/vertex_cache.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+namespace VertexCache {
+
+namespace {
+constexpr size_t cache_max_size = 256 * 256;
+// testing has proven that this size produces few misses in most games.
+// in total the cache size is 64Kb
+constexpr size_t cache_size = 512;
+
+// this is used to find if a vertex is in the cache
+static std::array<bool, cache_max_size> index_table = {};
```
Comment (wwylele) [inline src/video_core/vertex_cache.h]: This should go to a cpp file

File: src/video_core/vertex_cache.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+#ifdef _DEBUG
+static std::array<bool, cache_max_size> seen_table = {};
```
Comment (wwylele) [inline src/video_core/vertex_cache.h]: ditto, won't point out more cases

File: src/video_core/vertex_cache.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    index_table[position] = true;
+    cache[remap_index] = vertex;
+    max_index = std::max(max_index, remap_index);
+    min_index = std::min(min_index, remap_index);
+#ifdef _DEBUG
+    seen_table[position] = true;
+#endif
+}
+
+void clear() {
```
Comment (wwylele) [inline src/video_core/vertex_cache.h]: inline this function or make it to a cpp file

File: src/video_core/vertex_cache.h (hunk 2 above)
Comment (FernandoS27) [inline src/video_core/vertex_cache.h]: yeah, PICA has a vertex cache which is configured with GPUREG_POST_VERTEX_CACHE_NUM(0x2DD)  and is cleaned by GPUREG_VTX_FUNC(0x231).

**Comment by mailwl** [comment]
why do you close PRs? just commit into same branch

**Comment by wwylele** [comment]
@mailwl From IRC:

> Blinkhaw: I'm sorry but I'm closing all my PRs, tired of this shit. I won't contribute any longer except maybe on dynarmic

(I hope you can be often around in IRC #citra channel...)

**Comment by FernandoS27** [comment]
@mailwl I just grew tired having to deal with some of the mantainers. I prefer closing them than wasting anymore of my free time.

Don't get me wrong, there's very nice people among the mantainers (most of them) but there's also a few rotten apples that have no respect for other people's work.

So at the end, I decided the best for me is not to waste anymore of my time.

**Comment by tabnk** [comment]
:(

**Comment by tabnk** [comment]
Well, actually your **Texture overhaul** and **Vertex Cache** PR helped me to be able to play Super Mario 3D Land, Mario Kart 7, Super Mario Bro 2, Dragon Quest VII, The Legend of Zelda - A Link Between Worlds, etc at **60 fps full speed** without upgrading my current hardware spec.  

Your contribution will not be wasted. Thank a lot :)

**Comment by FernandoS27** [comment]
@tabnk thanks for the complements. If you are really really interested in playing games at "full speed" I got a hacky branch called autoskip. I never had any intention to do a PR for it since it's a hack that kinda skips frames when the emulated gpu can't keep up with the rest of the hardware. It should give you a good speed bust on many games (specialy Luigi's Mansion: Dark Moon). There's also a fix for avx compiling.

You can find all my branches here: https://github.com/FernandoS27/citra/branches

for best local build performance I suggest you use "-O3 -mtune=native -march=native"

**Comment by tabnk** [comment]
Nice. Will try it. Thank :D

**Comment by Gamedev1909** [comment]
https://discord.gg/Cm6Fufe Fernando you are always welcome in our discord we love what you are doing as progress is always welcome and should be no matter what and are right there is a lot of bad apples that really bring this project down it's why i left as well

**Pull request closed without merging**